File: backend/app/routers/report.py
"""Student progress report endpoint."""
from __future__ import annotations

from datetime import date, datetime, timezone

from fastapi import APIRouter, HTTPException

from app.config import get_curriculum
from app.db import get_conn

router = APIRouter(prefix="/students", tags=["report"])


@router.get("/{student_id}/report")
def get_report(student_id: str) -> dict:
    """Aggregate all progress data for the in-app report page."""
    with get_conn() as conn:
        student = conn.execute(
            "SELECT id, label FROM students WHERE id=?", (student_id,)
        ).fetchone()
        if not student:
            raise HTTPException(status_code=404, detail="Student not found.")

        buckets = conn.execute(
            "SELECT subtopic, bucket, updated_at FROM buckets WHERE student_id=?",
            (student_id,),
        ).fetchall()

        notes = conn.execute(
            """SELECT sn.student_breakthrough, sn.struggled_with, sn.topic_covered,
                      s.subtopic, s.started_at
               FROM session_notes sn
               JOIN sessions s ON sn.session_id = s.id
               WHERE s.student_id=?
               ORDER BY s.started_at DESC LIMIT 5""",
            (student_id,),
        ).fetchall()

        error_patterns = conn.execute(
            "SELECT subtopic, error_type, count FROM error_patterns "
            "WHERE student_id=? ORDER BY count DESC LIMIT 8",
            (student_id,),
        ).fetchall()

        quiz_attempts = conn.execute(
            """SELECT qa.score, qa.total, qa.submitted_at,
                      q.subtopic, q.sub_subtopic_label
               FROM quiz_attempts qa
               JOIN quizzes q ON qa.quiz_id = q.id
               WHERE qa.student_id=?
               ORDER BY qa.submitted_at DESC LIMIT 10""",
            (student_id,),
        ).fetchall()

        sessions = conn.execute(
            "SELECT started_at FROM sessions WHERE student_id=? ORDER BY started_at",
            (student_id,),
        ).fetchall()

        today_iso = date.today().isoformat()
        due_reviews = conn.execute(
            "SELECT subtopic, next_review FROM spaced_repetition "
            "WHERE student_id=? AND next_review <= ?",
            (student_id, today_iso),
        ).fetchall()

        mastery_rows = conn.execute(
            "SELECT concept_id, subtopic, mastery FROM concept_mastery WHERE student_id=?",
            (student_id,),
        ).fetchall()

    # Streak calculation
    session_dates = sorted({s["started_at"][:10] for s in sessions}, reverse=True)
    streak = 0
    if session_dates:
        prev = date.today().isoformat()
        for d in session_dates:
            if d == prev or (
                datetime.fromisoformat(prev).toordinal()
                - datetime.fromisoformat(d).toordinal() == 1
            ):
                streak += 1
                prev = d
            else:
                break

    curriculum = get_curriculum()
    subtopic_labels = {k: v.get("label", k) for k, v in curriculum.get("subtopics", {}).items()}

    # Mastery summary counts per subtopic
    mastery_summary: dict[str, dict[str, int]] = {}
    for m in mastery_rows:
        st = m["subtopic"]
        if st not in mastery_summary:
            mastery_summary[st] = {"solid": 0, "shaky": 0, "not_tested": 0}
        mastery_summary[st][m["mastery"]] = mastery_summary[st].get(m["mastery"], 0) + 1

    return {
        "student_id": student_id,
        "label": student["label"],
        "generated_at": today_iso,
        "streak_days": streak,
        "total_sessions": len(sessions),
        "total_quizzes": len(quiz_attempts),
        "buckets": [
            {
                "subtopic": r["subtopic"],
                "label": subtopic_labels.get(r["subtopic"], r["subtopic"]),
                "bucket": r["bucket"],
                "updated_at": r["updated_at"][:10],
            }
            for r in buckets
        ],
        "recent_notes": [
            {
                "breakthrough": n["student_breakthrough"],
                "struggled_with": n["struggled_with"],
                "topic": n["topic_covered"],
                "subtopic": n["subtopic"],
                "date": n["started_at"][:10],
            }
            for n in notes
            if n["student_breakthrough"]
        ],
        "error_patterns": [
            {
                "subtopic": e["subtopic"],
                "label": subtopic_labels.get(e["subtopic"], e["subtopic"]),
                "error_type": e["error_type"].replace("_", " "),
                "count": e["count"],
            }
            for e in error_patterns
        ],
        "recent_quizzes": [
            {
                "chapter": q["sub_subtopic_label"],
                "subtopic": q["subtopic"],
                "score": q["score"],
                "total": q["total"],
                "pct": round(q["score"] / q["total"] * 100) if q["total"] else 0,
                "date": q["submitted_at"][:10],
            }
            for q in quiz_attempts
        ],
        "due_reviews": [
            {
                "subtopic": r["subtopic"],
                "label": subtopic_labels.get(r["subtopic"], r["subtopic"]),
            }
            for r in due_reviews
        ],
        "mastery_summary": [
            {
                "subtopic": st,
                "label": subtopic_labels.get(st, st),
                **counts,
            }
            for st, counts in mastery_summary.items()
        ],
    }

File: frontend/js/report.js
// ── Progress Report Page ──────────────────────────────────────

const BUCKET_LABEL = { A: "Level A — Strong", B: "Level B — Developing", C: "Level C — Foundational" };
const BUCKET_CLASS = { A: "bucket-A", B: "bucket-B", C: "bucket-C" };

function el(id) { return document.getElementById(id); }

function renderBuckets(buckets) {
  const container = el("rpt-buckets");
  if (!buckets || buckets.length === 0) {
    container.innerHTML = '<p style="color:var(--text-lo);font-size:.82rem;">No diagnostic taken yet.</p>';
    return;
  }
  container.innerHTML = buckets.map(b => `
    <div class="rpt-bucket-card glass-card">
      <div class="rpt-bucket-name">${b.label}</div>
      <span class="bucket-badge ${BUCKET_CLASS[b.bucket] || ''}">${b.bucket}</span>
      <div class="rpt-bucket-date">assessed ${b.updated_at}</div>
    </div>
  `).join("");
}

function renderNotes(notes) {
  const section = el("rpt-notes-section");
  const container = el("rpt-notes");
  if (!notes || notes.length === 0) {
    section.style.display = "none";
    return;
  }
  container.innerHTML = notes.map(n => `
    <div class="glass-card rpt-note-card">
      <div class="rpt-note-meta">
        <span class="rpt-note-sub">${n.subtopic}</span>
        <span class="rpt-note-date">${n.date}</span>
      </div>
      <p class="rpt-note-text">${n.breakthrough}</p>
      ${n.struggled_with ? `<p class="rpt-note-struggled">Struggled with: ${n.struggled_with}</p>` : ""}
    </div>
  `).join("");
}

function renderMastery(summary) {
  const section = el("rpt-mastery-section");
  const container = el("rpt-mastery");
  if (!summary || summary.length === 0) {
    section.style.display = "none";
    return;
  }
  container.innerHTML = summary.map(s => {
    const total = (s.solid || 0) + (s.shaky || 0) + (s.not_tested || 0);
    const solidPct  = total ? Math.round((s.solid || 0) / total * 100) : 0;
    const shakyPct  = total ? Math.round((s.shaky || 0) / total * 100) : 0;
    return `
      <div class="rpt-mastery-row">
        <span class="rpt-mastery-label">${s.label}</span>
        <div class="rpt-mastery-bar-wrap">
          <div class="rpt-mastery-bar">
            <div class="rpt-bar-solid" style="width:${solidPct}%"></div>
            <div class="rpt-bar-shaky" style="width:${shakyPct}%"></div>
          </div>
          <span class="rpt-mastery-counts">
            <span class="rpt-count-solid">${s.solid || 0} solid</span>
            <span class="rpt-count-shaky">${s.shaky || 0} shaky</span>
            <span class="rpt-count-nt">${s.not_tested || 0} not tested</span>
          </span>
        </div>
      </div>`;
  }).join("");
}

function renderErrors(patterns) {
  const section = el("rpt-errors-section");
  const container = el("rpt-errors");
  if (!patterns || patterns.length === 0) {
    section.style.display = "none";
    return;
  }
  const max = Math.max(...patterns.map(p => p.count));
  container.innerHTML = patterns.map(p => {
    const pct = Math.round(p.count / max * 100);
    return `
      <div class="rpt-error-row">
        <div class="rpt-error-info">
          <span class="rpt-error-type">${p.error_type}</span>
          <span class="rpt-error-sub">${p.label}</span>
        </div>
        <div class="rpt-error-bar-wrap">
          <div class="rpt-error-bar" style="width:${pct}%"></div>
          <span class="rpt-error-count">${p.count}×</span>
        </div>
      </div>`;
  }).join("");
}

function renderQuizzes(quizzes) {
  const section = el("rpt-quiz-section");
  const container = el("rpt-quizzes");
  if (!quizzes || quizzes.length === 0) {
    section.style.display = "none";
    return;
  }
  container.innerHTML = quizzes.map(q => {
    const cls = q.pct >= 80 ? "quiz-good" : q.pct >= 50 ? "quiz-ok" : "quiz-bad";
    return `
      <div class="rpt-quiz-row">
        <div class="rpt-quiz-info">
          <span class="rpt-quiz-chapter">${q.chapter}</span>
          <span class="rpt-quiz-date">${q.date}</span>
        </div>
        <span class="rpt-quiz-score ${cls}">${q.score}/${q.total} (${q.pct}%)</span>
      </div>`;
  }).join("");
}

function renderDueReviews(due) {
  const section = el("rpt-due-section");
  if (!due || due.length === 0) return;
  el("rpt-due-list").innerHTML = due.map(r =>
    `<a href="chat.html?subtopic=${r.subtopic}" class="due-chip">${r.label} →</a>`
  ).join(" ");
  section.style.display = "flex";
}

function renderAchievements(achievements) {
  const section = el("rpt-achievements-section");
  const container = el("rpt-achievements");
  if (!achievements || achievements.length === 0) {
    section.style.display = "none";
    return;
  }
  container.innerHTML = achievements.map(a => `
    <div class="rpt-achievement-card ${a.earned ? 'earned' : 'locked'}">
      <div class="rpt-ach-icon">${a.icon}</div>
      <div class="rpt-ach-info">
        <div class="rpt-ach-title">${a.title}</div>
        <div class="rpt-ach-desc">${a.desc}</div>
        ${a.earned && a.unlocked_at ? `<div class="rpt-ach-date">${a.unlocked_at.slice(0,10)}</div>` : ""}
      </div>
      ${!a.earned ? '<div class="rpt-ach-lock">🔒</div>' : '<div class="rpt-ach-check">✓</div>'}
    </div>
  `).join("");
}

async function loadReport() {
  if (!Store.studentId) {
    el("report-loading").innerHTML =
      '<p style="color:var(--text-mid);">Please <a href="index.html" style="color:var(--cyan);">sign in</a> first.</p>';
    return;
  }

  const headerLbl = el("header-label");
  if (headerLbl) headerLbl.textContent = Store.studentLabel || "";

  try {
    const [data, achievements] = await Promise.all([
      Api.getReport(Store.studentId),
      Api.getAchievements(Store.studentId).catch(() => []),
    ]);

    el("report-loading").style.display = "none";
    el("report-content").style.display = "block";
    el("report-date").textContent = "Generated " + data.generated_at;

    el("stat-streak").textContent   = data.streak_days;
    el("stat-sessions").textContent = data.total_sessions;
    el("stat-quizzes").textContent  = data.total_quizzes;

    renderDueReviews(data.due_reviews);
    renderBuckets(data.buckets);
    renderNotes(data.recent_notes);
    renderMastery(data.mastery_summary);
    renderErrors(data.error_patterns);
    renderQuizzes(data.recent_quizzes);
    renderAchievements(achievements);
  } catch (e) {
    el("report-loading").textContent = "Could not load report.";
    console.error(e);
  }
}

loadReport();
